src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/OpenSslX509CertificateReader.cs

```diff
@@ -0,0 +1,572 @@
+
+        public string Issuer
+        {
+            get
+            {
+                if (_issuer == null)
+                {
+                    // IssuerName is mutable to callers in X509Certificate. We want to be
+                    // able to get the issuer even if IssuerName has been mutated, so we
+                    // don't use it here.
+                    _issuer = Interop.Crypto.LoadX500Name(Interop.Crypto.X509GetIssuerName(_cert)).Name;
+                }
+
+                return _issuer;
+            }
+        }
+
+        public string Subject
+        {
+            get
+            {
+                if (_subject == null)
+                {
+                    // SubjectName is mutable to callers in X509Certificate. We want to be
+                    // able to get the subject even if SubjectName has been mutated, so we
+                    // don't use it here.
+                    _subject = Interop.Crypto.LoadX500Name(Interop.Crypto.X509GetSubjectName(_cert)).Name;
+                }
+
+                return _subject;
+            }
+        }
+
+        public string LegacyIssuer => IssuerName.Decode(X500DistinguishedNameFlags.None);
+
+        public string LegacySubject => SubjectName.Decode(X500DistinguishedNameFlags.None);
+
+        public byte[] Thumbprint => Interop.Crypto.GetX509Thumbprint(_cert);
+
+        public string KeyAlgorithm => new Oid(Interop.AndroidCrypto.GetX509PublicKeyAlgorithm(_cert)).Value!;
+
+        public byte[] KeyAlgorithmParameters => Interop.Crypto.GetX509PublicKeyParameterBytes(_cert);
+
+        public byte[] PublicKeyValue
+        {
+            get
+            {
+                // AndroidCrypto returns the SubjectPublicKeyInfo - extract just the SubjectPublicKey
+                byte[] bytes = Interop.AndroidCrypto.GetX509PublicKeyBytes(_cert);
+                return SubjectPublicKeyInfoAsn.Decode(bytes, AsnEncodingRules.DER).SubjectPublicKey.ToArray();
+            }
+        }
+
+        public byte[] SerialNumber => Interop.AndroidCrypto.X509GetSerialNumber(_cert);
+
+        public string SignatureAlgorithm => Interop.AndroidCrypto.GetX509SignatureAlgorithm(_cert);
+
+        public DateTime NotAfter
+        {
+            get
+            {
+                ulong msFromUnixEpoch = Interop.AndroidCrypto.GetX509NotAfter(_cert);
+                return DateTime.UnixEpoch.AddMilliseconds(msFromUnixEpoch).ToLocalTime();
+            }
+        }
+
+        public DateTime NotBefore
+        {
+            get
+            {
+                ulong msFromUnixEpoch = Interop.AndroidCrypto.GetX509NotBefore(_cert);
+                return DateTime.UnixEpoch.AddMilliseconds(msFromUnixEpoch).ToLocalTime();
+            }
+        }
+
+        public byte[] RawData => Interop.AndroidCrypto.EncodeX509(_cert);
+
+        public int Version
+        {
+            get
+            {
+                int version = Interop.Crypto.GetX509Version(_cert);
```

Comment:
I'm finding the mix of Crypto and AndroidCrypto hard to reason about.  Can you just make all the Android-needed routines be in AndroidCrypto and stop including the one that assumes OpenSSL?

Reply:
I moved these to be in `AndroidCrypto`. The only two left are `X509Destroy` and `X509UpRef`, since those are used by code shared with the Unix build (for now) - will plan to move those as we fork off more.